app/controllers/api/storage.php

```diff
@@ -245,6 +246,218 @@
+        /** @var Appwrite\Utopia\Response $response */
+        /** @var Utopia\Database\Database $dbForInternal */
+        /** @var Appwrite\Database\Document $user */
+        /** @var Appwrite\Event\Event $audits */
+        /** @var Appwrite\Event\Event $usage */
+
+        $bucket = $dbForInternal->getDocument('buckets',$bucketId);
+
+        if($bucket->isEmpty()) {
+            throw new Exception("Unable to find the bucket", 404);
+        }
+
+        $file = $request->getFiles('file');
+
+        /*
+         * Validators
+         */
+        // $fileType = new FileType($bucket->getAttribute('allowedFileExtensions', ['*']));
+        $fileSize = new FileSize($bucket->getAttribute('maximumFileSize', 0));
+        $upload = new Upload();
+
+        if (empty($file)) {
+            throw new Exception('No file sent', 400);
+        }
+
+        // Make sure we handle a single file and multiple files the same way
+        $file['name'] = (\is_array($file['name']) && isset($file['name'][0])) ? $file['name'][0] : $file['name'];
+        $file['tmp_name'] = (\is_array($file['tmp_name']) && isset($file['tmp_name'][0])) ? $file['tmp_name'][0] : $file['tmp_name'];
+        $file['size'] = (\is_array($file['size']) && isset($file['size'][0])) ? $file['size'][0] : $file['size'];
+
+        // Check if file type is allowed (feature for project settings?)
+        // if (!$fileType->isValid($file['tmp_name'])) {
+        // throw new Exception('File type not allowed', 400);
+        // }
+
+        if (!$fileSize->isValid($file['size'])) { // Check if file size is exceeding allowed limit
+            throw new Exception('File size not allowed', 400);
+        }
+
+        $device = Storage::getDevice('files');
+
+        if (!$upload->isValid($file['tmp_name'])) {
+            throw new Exception('Invalid file', 403);
+        }
+
+        // Save to storage
+        $size = $device->getFileSize($file['tmp_name']);
+        $path = $device->getPath(\uniqid().'.'.\pathinfo($file['name'], PATHINFO_EXTENSION));
+        $path = $bucket->getId() . '/' . $path;
+        
+        if (!$device->upload($file['tmp_name'], $path)) { // TODO deprecate 'upload' and replace with 'move'
+            throw new Exception('Failed moving file', 500);
+        }
+
+        $mimeType = $device->getFileMimeType($path); // Get mime-type before compression and encryption
+
+        if (App::getEnv('_APP_STORAGE_ANTIVIRUS') === 'enabled' && $bucket->getAttribute('antiVirus', true) && $size <= APP_LIMIT_ANTIVIRUS) {
+            $antiVirus = new Network(App::getEnv('_APP_STORAGE_ANTIVIRUS_HOST', 'clamav'),
+                (int) App::getEnv('_APP_STORAGE_ANTIVIRUS_PORT', 3310));
+
+            if (!$antiVirus->fileScan($path)) {
+                $device->delete($path);
+                throw new Exception('Invalid file', 403);
+            }
+        }
+
+        // Compression
+        $data = $device->read($path);
+        if($size <= APP_LIMIT_COMPRESSION) {
```

Comment:
Would be important to add a test that validates a bigger file is uploaded properly and read properly (compression + decompression both happen or not)

Reply:
done!